### PR TITLE
Media: Add state behaviors for series upload

### PR DESCRIPTION
**Pull request by aduth**
~~Blocks #9675~~
Related: #5046

This pull request seeks to start implementation of media Redux state handlers. The included behaviors are quite basic, including only uploads of one or many files, without validation or compatibility with prior media Flux state.

__Implementation Notes:__

Recognizing a common pattern of keying reducer state objects by site ID, included in the changes is a proposed `createSiteReducer`, an enhanced variant of `createReducer` which automatically keys the state returned from the reducer by an assumed `siteId` property on the action.

__Testing Instructions:__

Ensure that Mocha tests pass:

```
npm run test-client
```

The commit 6922b25 is a temporary and unfinished demonstration of expected behaviors. It replaces existing Flux dispatching with the new Redux action creators usage. Verify that after uploading one or many files to the editor media library that the uploads occur in series order (i.e. uploads one at a time). This is easiest to verify using the [Redux DevTools extension](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en), but you can also use the newly introduced [action logging feature](https://github.com/Automattic/wp-calypso/tree/master/client/state/console-dispatch).

upload

1. Navigate to the [post editor](http://calypso.localhost:3000/post)
2. Select a site, if prompted
3. Click Add Media in the editor toolbar
4. Upload one or more media items
5. Observe that Redux actions are logged in expected sequence
6. Refresh the page (items will **not** show in library in same session yet)
7. Click Add Media in the editor toolbar
8. Note that items were uploaded successfully

cc @tyxla @gwwar

### CODE REVIEW

File: client/state/utils.js [hunk 1]
```diff
@@ -116,3 +116,47 @@ export function createReducer( initialState = null, customHandlers = {}, schema
...
+ *
+ * @param  {*}        initialSiteState Initial state per site ID
+ * @param  {Object}   customHandlers   Object mapping action types to state
+ *                                     action handlers
+ * @param  {?Object}  schema           JSON schema object for deserialization
+ *                                     validation
+ * @return {Function}                  Reducer function
+ */
+export function createSiteReducer( initialSiteState = null, customHandlers = {}, schema = null ) {
+	if ( schema ) {
```
Comment (gwwar) [inline client/state/utils.js]: Did you mean this as a falsey check?

Comment (gwwar) [inline client/state/utils.js]: Ah ic, perhaps give the schema below a different name?

**Comment by gwwar** [comment]
Nice @aduth ! This diff is a bit larger so I'll try to take a closer 👀 tomorrow

File: client/state/utils.js (hunk 1 above)
Comment (aduth) [inline client/state/utils.js]: I added some inline documentation to the function in 948bb5a. Does this help clarify?

Comment (gwwar) [inline client/state/utils.js]: Yes, thanks 👍 Just took me a bit longer to read out the schema part since it looked like a fallback.

File: client/state/utils.js [hunk 2]
```diff
@@ -116,3 +116,54 @@ export function createReducer( initialState = null, customHandlers = {}, schema
...
+ * @see createReducer
+ *
+ * @param  {*}        initialSiteState Initial state per site ID
+ * @param  {Object}   customHandlers   Object mapping action types to state
+ *                                     action handlers
+ * @param  {?Object}  schema           JSON schema object for deserialization
+ *                                     validation
+ * @return {Function}                  Reducer function
+ */
+export function createSiteReducer( initialSiteState = null, customHandlers = {}, schema = null ) {
```
Comment (gwwar) [inline client/state/utils.js]: We can leave this for another PR, but it'd be handy to explain what situations we use `createReducer` and `createSiteReducer` in [our-approach-to-data](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md). I just realized we use `createReducer` in our examples but never explain what it does.

File: client/state/test/utils.js [hunk 3]
```diff
@@ -235,4 +235,51 @@ describe( 'utils', () => {
...
+			const state = reducer( originalState, {
+				type: 'TEST_ADD',
+				siteId: 2916284,
+				value: 1
+			} );
+
+			expect( state ).to.equal( originalState );
+		} );
+
+		it( 'should augment schema to account for site keyed structure', () => {
```
Comment (gwwar) [inline client/state/test/utils.js]: Optional, but we can also test to see if we respect the passed in schema / also when the keys are not siteids

**Comment by gwwar** [comment]
Thanks @aduth this looks like a good first step. 👍 I verified that the Redux actions worked as expected.

Are we planning on dropping 6922b25 before merge?  As is the branch will introduce a number of regressions:
1. Client side validation + missing feedback for validation checks ( eg a user uploading video on a free account ).
2. No transient loaders are shown
3. List view doesn't refresh after successful upload
4. Weird behavior when dropping media directly to the editor

File: client/state/action-types.js [hunk 4]
```diff
@@ -194,7 +194,12 @@ export const KEYRING_SERVICES_REQUEST_SUCCESS = 'KEYRING_SERVICES_REQUEST_SUCCES
 export const LAYOUT_FOCUS_SET = 'LAYOUT_FOCUS_SET';
 export const LAYOUT_NEXT_FOCUS_SET = 'LAYOUT_NEXT_FOCUS_SET';
 export const LAYOUT_NEXT_FOCUS_ACTIVATE = 'LAYOUT_NEXT_FOCUS_ACTIVATE';
+export const MEDIA_ITEMS_RECEIVE = 'MEDIA_ITEMS_RECEIVE';
```
Comment (tyxla) [inline client/state/action-types.js]: While working on reduxifying shortcodes (#8857) I noticed that some of the functionality requires granular information about media items receiving (for example https://github.com/Automattic/wp-calypso/pull/8857/files#diff-ac2308bd4d1d8c2075ad658ea3b46b5bR55). To support that, in addition to `MEDIA_ITEMS_RECEIVE` we'll probably have to dispatch a `MEDIA_ITEM_RECEIVE` for each single media item. You can observe the current flux actions for that - `RECEIVE_MEDIA_ITEMS` and `RECEIVE_MEDIA_ITEM`.

File: client/state/utils.js (hunk 2 above)
Comment (aduth) [inline client/state/utils.js]: Yeah, and more generally we need better documentation of all of our state utilities.

File: client/state/test/utils.js (hunk 3 above)
Comment (aduth) [inline client/state/test/utils.js]: >Optional, but we can also test to see if we respect the passed in schema / also when the keys are not siteids

Can you clarify what you mean here? A test to see that an invalid state would be discarded?

**Comment by aduth** [comment]
>Are we planning on dropping 6922b25 before merge? As is the branch will introduce a number of regressions:

Definitely plan to drop, hence the commit name 😄

File: client/state/action-types.js (hunk 4 above)
Comment (aduth) [inline client/state/action-types.js]: >To support that, in addition to `MEDIA_ITEMS_RECEIVE` we'll probably have to dispatch a `MEDIA_ITEM_RECEIVE` for each single media item.

It's not clear from your comment, but what would a separate `MEDIA_ITEM_RECEIVE` action enable that couldn't be achieved by looping over items of `MEDIA_ITEMS_RECEIVE` ?

Ideally trying to avoid multiple actions to simplify tests and implementation of reducers. See sites [as an example](https://github.com/Automattic/wp-calypso/blob/207d4eda3c6a8b999cb270f5c63348cacbd34f64/client/state/current-user/reducer.js#L63-L84) of why this can be a pain.

File: client/state/test/utils.js (hunk 3 above)
Comment (gwwar) [inline client/state/test/utils.js]: Optional, but I thought it might be interesting to have these cases:
1. A test to verify that the siteId `patternProperties` works
2. A test to verify that the item schema is being respected. (For example passing through an object instead of an array).

Comment (aduth) [inline client/state/test/utils.js]: I'm still confused - isn't that what's already being tested here? The site-specific reducer returns an array, but I'm validating against the keyed object, and the schema provided in `before` checks that it matches the expected `patternProperties` for a site ID and that the shape of the inner array is as expected.

Comment (gwwar) [inline client/state/test/utils.js]: I mean for the rejection cases. For example imagine if we forgot to add `additonalProperties: false`

File: client/state/action-types.js (hunk 4 above)
Comment (tyxla) [inline client/state/action-types.js]: My point was that by dispatching a separate single item action we could avoid boilerplate code like this: https://github.com/Automattic/wp-calypso/pull/8857/files#diff-ac2308bd4d1d8c2075ad658ea3b46b5bR60.

But I think your point makes sense, more actions could overcomplex the implementations in some cases. Feel free to ignore my comment.

File: client/state/test/utils.js (hunk 3 above)
Comment (aduth) [inline client/state/test/utils.js]: Added two invalidation tests in 0d87a07.

File: client/state/action-types.js (hunk 4 above)
Comment (aduth) [inline client/state/action-types.js]: The problem in your link is that the action creator could dispatch with a single or multiple items. Here, we only have a single action type that will always include an array of items.

**Comment by gwwar** [comment]
Feel free to 🚢  when 6922b25 has been dropped. I can help test again for regressions if you'd like more 👀

**Comment by aduth** [comment]
I marked this ready to merge, but upon some consideration in #9675 I found a reasonable path forward with the current Flux implementation and am going to punt this for at least a few days until I or someone else can commit to fleshing out the Redux media implementation (particularly implementing list and querying behaviors).

**Comment by aduth** [comment]
Since I've found myself reimplementing the same reducers and logic in other pull requests, I decided to revisit this to try to get it out the door. I've rebased, dropped `createSiteReducer` in favor of @dmsnell 's `keyedReducer` (#9991) and tried using the API middleware to achieve the series upload queueing effect (6acbabd).

However, I'm running into an issue because of how the actions are dropped by the API middleware. Ideally in my case I can hook into the enqueueing action and trigger the upload if no other uploads are in progress. Because the middleware drops the actions, this is not possible because the action never reaches the reducer managing the queue. @dmsnell : Noting your [comment on duplication / prioritization](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/README.md#implementation), do you think this is something that should be achieved in other ways - like a piece of state that the handler checks? Or do you have any suggestions for how I might achieve the effect I'm looking for with the API middleware? Or should I revert back to the standalone middleware?

Also, just now realizing that I've only partially adopted the API middleware - `uploadFile` should be changed from a thunk to issuing its request in the media API handlers file.

**Comment by dmsnell** [comment]
> Because the middleware drops the actions, this is not possible because the action never reaches the reducer managing the queue.

Just last night I was thinking about this too @aduth and I think a reasonable method is to provide an escape hatch in action `meta`, that or pass an altered `dispatch` to the middleware which can be used to bypass more middleware.

```js
dispatch( fromMiddleware( updateSpline( splineId, points ) ) )

// action is…

{
	type: SPLINE_UPDATE,
	splineId,
	points,
	meta: {
		skipMiddleware: true,
	}
}
```

and then we check for this flag in the middleware at the top and pass it to `next()` if it's there. this is a very pragmatic solution and I'm not sure if it's the best, but it's pretty straightforward.

for actions unrelated to what we trapped (posts middleware wanting to issue get site actions for example) then we can simply issue requests to `dispatch` like usual

File: client/state/data-layer/wpcom/sites/media/index.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+export function uploadNext( store ) {
+	const nextQueued = getNextQueuedUpload( store.getState() );
+	if ( nextQueued ) {
+		const { siteId, file } = nextQueued;
+		store.dispatch( uploadFile( siteId, file ) );
+	}
+}
+
+export function maybeUploadFirst( store, action ) {
+	if ( ! isMediaUploadInProgress( store.getState() ) ) {
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/media/index.js]: Not sure if this is being used elsewhere, but if the value of whether or not we are currently uploading is really contained to this file, does it need to be stored in Redux?

If we kill browser session and reload, do we expect that value to persist and for the app to reinitialize at this point?

If not, I'm open to hiding network-layer-related data here in the network-layer-related files.

File: client/state/data-layer/wpcom/sites/media/index.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+		store.dispatch( uploadFile( action.siteId, action.files[ 0 ] ) );
+	}
+}
+
+export default {
+	[ MEDIA_FILE_UPLOADS_ENQUEUE ]: [ maybeUploadFirst ],
+	[ MEDIA_FILE_UPLOAD_FAILURE ]: [ uploadNext ],
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/media/index.js]: this would be a great place to inject some other responder such as `announceFailure` which could dispatch a `notice` with a simple message explaining the failure

```js
[ MEDIA_FILE_UPLOAD_FAILURE ]: [ announceFailure, uploadNext ]
```

File: client/state/media/actions.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			} );
+		} ).catch( () => {
+			dispatch( {
+				type: MEDIA_FILE_UPLOAD_FAILURE,
+				siteId,
+				file
+			} );
+		} );
+	};
+}
```
Comment (dmsnell) [inline client/state/media/actions.js]: You may have mentioned this in your general comment on the PR, but are you planning on introducing the API middleware and _not_ putting this API code in it?

trivially this can be cut and pasted there leaving a shell action here.

File: client/state/media/reducer.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export const uploadQueue = createReducer( [], {
+	[ MEDIA_FILE_UPLOADS_ENQUEUE ]: ( state, { siteId, files } ) => [
+		...state,
+		...files.map( ( file ) => [ siteId, file ] )
+	],
+	[ MEDIA_FILE_UPLOAD_FAILURE ]: ( state, { file } ) => reject( state, ( [ , queued ] ) => queued === file ),
+	[ MEDIA_FILE_UPLOAD_SUCCESS ]: ( state, { file } ) => reject( state, ( [ , queued ] ) => queued === file )
```
Comment (dmsnell) [inline client/state/media/reducer.js]: in some of my code I have found it helpful to include a comment in destructured lists indicating what's missing

```js
[ /* sprocket */, queued ] = …
```

File: client/state/media/reducer.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+			[ transientMedia.ID ]: transientMedia
+		};
+	},
+	[ MEDIA_FILE_UPLOADS_ENQUEUE ]: ( state, action ) => {
+		return {
+			...state,
+			...keyBy( map( action.files, createTransientMedia ), 'ID' )
+		};
+	}
+} ) );
```
Comment (dmsnell) [inline client/state/media/reducer.js]: Just taking notes here:
 - once attempted to be added, media items _never_ leave the state tree
 - we have a collection of collections keyed by `( siteId, mediaId )`

Could this be a flat structure? Would there be benefit only having one layer?

Is there a way to use `keyedReducer` again on the media `ID`?  the entire structure of the state inside a media item here is completely opaque. I wonder if we need to expand the `keyedReducer()` concept to handle large influxes of items like this or if we could get by turning a single dispatch with a list of items into a list of dispatches with single items. maybe some hybrid would help too.

**Comment by dmsnell** [comment]
@aduth this is a big PR but it's easy to follow - thanks 🙇 !

although I left these in the code, I have a couple of high-level questions and I'm not sure how tied in to the existing system we are, whether this is green-field territory or not.

**Comment by aduth** [comment]
>Just last night I was thinking about this too @aduth and I think a reasonable method is to provide an escape hatch in action `meta`, that or pass an altered `dispatch` to the middleware which can be used to bypass more middleware.

Rather than escape hatches, are we certain that dropping the action is the best default? Currently the arguments in favor seem a bit theoretical. Another option to consider is giving the handlers more of a middleware signature in being able to control `next` / `action`, requiring an explicit call to have the action not be dropped.

File: client/state/data-layer/wpcom/sites/media/index.js (hunk 5 above)
Comment (aduth) [inline client/state/data-layer/wpcom/sites/media/index.js]: >Not sure if this is being used elsewhere, but if the value of whether or not we are currently uploading is really contained to this file, does it need to be stored in Redux?

Looking at the current UI requirements, we'll need some upload indicators managed in state, but specifically for the individual items themselves. At the moment, we're not using the queue in any form. To answer your question, no, I don't expect any persistence between sessions, and this will only ever occur in the browser now. That said, I'm wary of module-level variables existing in state, given the SSR use-case. Seems unlikely that we'll need this for SSR anytime in the near future, but I don't want to set a potentially problematic precedent for others to follow.

File: client/state/data-layer/wpcom/sites/media/index.js (hunk 6 above)
Comment (aduth) [inline client/state/data-layer/wpcom/sites/media/index.js]: >this would be a great place to inject some other responder such as `announceFailure` which could dispatch a `notice` with a simple message explaining the failure

Indeed, we'll need something like this to recreate the current media UI in Redux. The Flux implementation defines a [`MediaValidationStore`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/media/validation-store.js) to track errors. To recreate this, a reducer following the failure actions could track these errors, though this brings us back to the issue of losing dropped actions. Along these lines, if we did want to continue dropping actions, it would seem the next best thing would be for the network handlers to be responsible for dispatching back into our store the equivalent actions those failures would incur?

File: client/state/media/reducer.js (hunk 8 above)
Comment (aduth) [inline client/state/media/reducer.js]: >in some of my code I have found it helpful to include a comment in destructured lists indicating what's missing

Yeah, makes sense. If ESLint doesn't complain about it, would it make sense to just include the destructured variable uncommented?

File: client/state/media/actions.js (hunk 7 above)
Comment (aduth) [inline client/state/media/actions.js]: >Yeah, I'll move this to the API middleware shortly.

Updated in b277f57.

File: client/state/media/reducer.js (hunk 9 above)
Comment (dmsnell) [inline client/state/media/reducer.js]: > We may also want to recreate the behavior of the Flux store in tracking "pointers" from transient media IDs to their successfully uploaded IDs.

I'm not sure if the Importer does it exactly this way off hand, but to a certain extent I feel like the temporary ids should be able to be replaced entirely from within the API code. Going back to "the local state is the global truth" concept, the code which actually uploads the media knows its transient ID and also gets the response from the server, where it could update/rename/replace that in the state.

File: client/state/data-layer/wpcom/sites/media/index.js (hunk 6 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/media/index.js]: I responded to this below in my reply to "escape hatches"

**Comment by dmsnell** [comment]
> Rather than escape hatches, are we certain that dropping the action is the best default? Currently the arguments in favor seem a bit theoretical. Another option to consider is giving the handlers more of a middleware signature in being able to control next / action, requiring an explicit call to have the action not be dropped.

I'm definitely not certain that dropping is the best default. On the positive side it simplifies the API handling interface. Additionally, it was built with the goals of allowing for gradual replacement of data sources (WP-API vs WPCOM API) though I'm not sure if that's even a pressing need of ours.

My original hope was that we could have actions that explicitly meant, "I need this data" and "I need to update this data," but then I reused our `_REQUEST` actions and that could have made this worse.

Adding `next` seems reasonable.

Do you think there would be many more cases where we would want to pass the action through?

---

Further, this might bring us back to the question of action naming. If we reused the requester action name with included `meta` indicating fetch status then `REQUEST_FAILED` would have more meaning than "here is the response data from the failed fetch from the server."

Maybe by default we just reuse the requesting action in the API middleware and allow for reuse of the `_FAILURE` and `_SUCCESS` actions to be the things handled by the reducers _if indeed they need to store state specifically related to the success of the fetch_ such as spinners and indicators.

Could this solution also remove the need for action pass-through?

File: client/state/data-layer/wpcom/sites/media/test/index.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+			.withArgs( sinon.match.func ).callsArgWith( 0, store.dispatch );
+		sandbox.stub( store, 'getState', () => state );
+	} );
+
+	function stubStateUploadsInProgress( uploads = 0 ) {
+		state = {
+			media: {
+				uploadQueue: times( uploads, () => [ 2916284, 'https://wordpress.com/i/stats-icon.gif' ] )
+			}
+		};
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/media/test/index.js]: This kind of state mocking is why I have tried to start writing composed selectors. On the negative side, we get more function call overhead, but on the positive side, we get easy mocking and more composable state code.

For an example see [the AT-AT selectors](https://github.com/Automattic/wp-calypso/blob/f69d6577bb6c5232d993365b4d440f42c8b6a254/client/state/automated-transfer/selectors.js#L47)

**Comment by aduth** [comment]
@dmsnell I'm not really clear on the alternatives you're proposing. Considering the example of having state track request progress for the purposes of showing a spinner, how would that be implemented? Presumably the reducer would need to listen for the initial `FOO_REQUEST` action to start considering a request in progress, but that would not be possible when the actions are dropped by the API middleware.

**Comment by dmsnell** [comment]
> Considering the example of having state track request progress for the purposes of showing a spinner, how would that be implemented?

it's worth prefixing this here with a big 'ole "I don't know" and "I'm trying to answer the same ❓ myself", so hopefully this sets the tone as non-argumentative.

I would want to start with the question, why do we even have a spinner? if this were being saved as a local file we probably wouldn't even have one. On the other hand, Calypso _lives_ in the web and long transfers are part-and-parcel of uploading.

In one of your PRs I think we were "tricking" the UI into thinking that the image was already uploaded and continuing the process in the background, maybe sending a notification once the upload finished or failed. This is more or less how I view what's going on here.

in fact, it seems reasonable (and common) to add a new progress bar to a stack of "current uploads" as rich notification bubbles inside of Calypso. in this case, each upload would have its own progress state with either a percentage or current/total count.

could we envision a top-level global uploads state? maybe instead of reducing over the media objects, we might create a new upload object to track; something which could definitely be dispatched from inside the API layer.

```js
const uploadReducer = combineReducer( {
	bytesLoaded,
	bytesTotal,
	failureTYpe
} );

const uploads = keyedReducer( 'uploadId', uploadReducer );
```

now when we want to upload a media item, we issue

```js
const uploadImage = ( { dispatch }, action ) => {
	const uploadId = uniq();

	dispatch( { type: UPLOAD_UPDATE, uploadId, bytesTotal: action.file.size } );
	const req = wpcom.req.post( … );

	req.upload.onprogress = ( { loaded } ) =>
		dispatch( { 
			type: UPLOAD_UPDATE, 
			uploadId,
			bytesLoaded: loaded
		} )
}
```

this is more complicated than a boolean `isUploading` but either could do. the 👿 is in the details, right? maybe there could be benefits here and in other places, however, to having _uploads_ be a separate thing.

if I have failed to address the question of "how would this prevent our need to pass through the actions" please do ask me to clarify or challenge my response :smile:

**Comment by aduth** [comment]
@dmsnell I think your idea seems wise for media, even if not a near-term possibility (I recall trying progress updates before and it's not as simple as it sounds; I suspect due to the proxy frame?). In any case, the idea seems to boil down to the handler dispatching a separate action to indicate progress. But while this seems a good fit for media, it doesn't address the much more common case of retrieving some data from the API and showing some placeholders while that data is being fetched. This pattern is near-universal and relies on state knowing when the request is initiated and when it completes. Sure, we could dispatch different actions, maybe `FOO_PROGRESS`, from our `FOO_REQUEST` handlers, but... that seems like alot of actions? Or maybe not, if we consolidate reducer logic for `FOO_REQUEST`, `FOO_REQUEST_SUCCESS`, and `FOO_REQUEST_FAILURE` into this one action payload.

**Comment by dmsnell** [comment]
> (I recall trying progress updates before and it's not as simple as it sounds; I suspect due to the proxy frame?)

for progress updates in general we have had no problems in the importer to my knowledge and that's been in for a year and a half now

> it doesn't address the much more common case of retrieving some data from the API and showing some placeholders while that data is being fetched.

yes indeed. this is a bigger problem. and I think I get what you are saying now about passing the actions through. I think I'm a bit blinded to it because I try to imagine something like `null` meaning "we are retrieving data" while things like empty strings, empty lists, empty objects imply "data has been retrieved and there is nothing to show."

I'm going to have to consider this more while we think about making pragmatic changes here. ideally I think I like using `null` states but would prefer the kinds of data types other languages provide with pattern matching, not that it matters for Calypso.

```purescript
{-
State is either unknown because we haven't yet
fetched it from the server, thus "loading" or it
has been successfully loaded and therefore we not
only have the data itself but also an indication of how
long ago it was that we updated it
-}
data Uploads
    = Unknown
    | Uploads Datetime (Array UploadData)

render :: Uploads -> Html
render Unknown = showPlaceholderList
render (Uploads lastUpdate uploads) = showUploadList uploads
```

I have seen these kinds of things being done in JavaScript through the use of constructor functions but I've never been happy with them.

**Comment by dmsnell** [comment]
@aduth much has changed since this PR was developed. we didn't have someone already address these issues in another PR did we?

anyway, I'm sure the data layer would bring implications for this code since it's now fully in place.

just checking on the status and future here. thanks!

**Comment by aduth** [comment]
Thanks for the nudge @dmsnell . This is very out of date indeed. I don't think all of the changes have been satisfied elsewhere, but this one stalled because there wasn't a clear path to converting all existing usage to the new recommendations, and I worried about letting unused code linger (similar conundrum as with `sites-list`).

Let's close for now and revisit as reference later.
